test/test_jit.py

```diff
@@ -3984,6 +3984,97 @@ def test_while(a, b):
                 return a + b
             ''')
 
+    def test_optional_refinement(self):
+        @torch.jit.script
+        def test_if_none_assignment(x):
+            # type: (Optional[int]) -> int
+            if x is None:
+                x = 1
+            return x + 1
+
+        @torch.jit.script
+        def test_ternary(x):
+            # type: (Optional[int]) -> int
+            x = 1 if x is None else 2
+            return x
+
+        @torch.jit.script
+        def test_not_none(x):
+            # type: (Optional[int])
```

Comment:
The `None` return [should be explicit](https://mypy.readthedocs.io/en/latest/kinds_of_types.html#the-any-type), there were some mypy issues at some point over this same thing